io/src/lib.rs

```diff
@@ -18,30 +18,224 @@
+    }
+    #[cfg(all(feature = "alloc", not(feature = "std")))]
+    impl IntoBoxDynDebug for &str {
+        fn into(self) -> Box<dyn Debug + Send + Sync + 'static> {
+            Box::new(String::from(self))
+        }
+    }
+    #[cfg(all(feature = "alloc", not(feature = "std")))]
+    impl IntoBoxDynDebug for String {
+        fn into(self) -> Box<dyn Debug + Send + Sync + 'static> {
+            Box::new(self)
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct Error {
+        kind: ErrorKind,
+
+        #[cfg(feature = "std")]
+        error: Option<Box<dyn std::error::Error + Send + Sync + 'static>>,
+        #[cfg(all(feature = "alloc", not(feature = "std")))]
+        error: Option<Box<dyn Debug + Send + Sync + 'static>>,
+    }
+    impl Error {
+        #[cfg(feature = "std")]
+        pub fn new<E>(kind: ErrorKind, error: E) -> Error where
+            E: Into<Box<dyn std::error::Error + Send + Sync + 'static>>
+        {
+            Self { kind, error: Some(error.into()) }
+        }
+        #[cfg(all(feature = "alloc", not(feature = "std")))]
+        pub fn new<E: IntoBoxDynDebug>(kind: ErrorKind, error: E) -> Error {
+            Self { kind, error: Some(error.into()) }
+        }
+
+        pub fn kind(&self) -> ErrorKind {
+            self.kind
+        }
+    }
+
+    impl From<ErrorKind> for Error {
+        fn from(kind: ErrorKind) -> Error {
+            Self {
+                kind,
+                #[cfg(any(feature = "std", feature = "alloc"))]
+                error: None,
+            }
+        }
+    }
+
+    impl Display for Error {
+        fn fmt(&self, fmt: &mut Formatter) -> core::result::Result<(), core::fmt::Error> {
+            fmt.write_fmt(format_args!("I/O Error: {}", self.kind.description()))?;
+            #[cfg(any(feature = "alloc", feature = "std"))]
+            if let Some(e) = &self.error {
+                fmt.write_fmt(format_args!(". {:?}", e))?;
+            }
+            Ok(())
+        }
+    }
+
+    #[cfg(feature = "std")]
+    impl std::error::Error for Error {
+        fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+            self.error.as_ref().and_then(|e| e.as_ref().source())
+        }
+        #[allow(deprecated)]
+        fn description(&self) -> &str {
+            match self.error.as_ref() {
+                Some(e) => e.description(),
+                None => self.kind.description()
+            }
+        }
+        #[allow(deprecated)]
+        fn cause(&self) -> Option<&dyn std::error::Error> {
+            self.error.as_ref().and_then(|e| e.as_ref().cause())
+        }
+    }
 
-    #[cfg(all(not(feature = "std"), not(feature = "core2")))]
-    compile_error!("At least one of std or core2 must be enabled");
+    impl Error {
+        #[cfg(feature = "std")]
+        pub fn get_ref(&self) -> Option<&(dyn std::error::Error + Send + Sync + 'static)> {
+            #[allow(clippy::option_as_ref_deref)] // Doesn't yet meet MSRV
```

Comment:
In baf0fc26c198d9d238661cfa2a4b09febb0bfdfe:

You can create a clippy.toml file with a `msrv` field and it should stop making suggestions that aren't allowed with our MSRV.

Reply:
Because it *does* meet MSRV which is 1.48 and `as_deref` was added in 1.42. We can use it.